File: packages/bot/src/lib/mcts-types.ts
// services/mcts-types.ts
import { Card, Pile, Rank } from '@vinto/shapes';
import { BotMemory } from './bot-memory';

/**
 * Type of move in the game
 */
export type MCTSMoveType =
  | 'draw'
  | 'take-discard'
  | 'use-action'
  | 'swap'
  | 'discard'
  | 'toss-in'
  | 'call-vinto'
  | 'pass';

/**
 * Target for an action
 */
export interface MCTSActionTarget {
  playerId: string;
  position: number;
}

/**
 * A move in the game
 */
export interface MCTSMove {
  type: MCTSMoveType;
  playerId: string;

  // Action-specific data
  actionCard?: Card;
  targets?: MCTSActionTarget[];
  swapPosition?: number;
  declaredRank?: Rank;
  tossInPositions?: number[]; // Array of positions for multi-card toss-in
  shouldSwap?: boolean; // For peek-and-swap actions
}

/**
 * Simplified player state for MCTS
 */
export interface MCTSPlayerState {
  id: string;
  cardCount: number;
  knownCards: Map<number, { card: Card | null; confidence: number }>; // Position -> Memory (from bot's perspective)
  score: number; // Estimated score based on known cards
}

/**
 * Game state for MCTS simulation
 */
export interface MCTSGameState {
  // Known information
  players: MCTSPlayerState[];
  currentPlayerIndex: number;
  botPlayerId: string;

  discardPileTop: Card | null;
  discardPile: Pile; // Full discard pile history
  deckSize: number;

  // Bot's belief state (imperfect information)
  botMemory: BotMemory;

  // Hidden information (sampled during determinization)
  // Map of playerId-position -> Card
  hiddenCards: Map<string, Card>;

  // Current turn state
  pendingCard: Card | null; // Card drawn but not yet swapped/discarded
  isTossInPhase: boolean;
  tossInRanks?: [Rank, ...Rank[]]; // Available ranks for toss-in (multiple ranks can be valid)

  // Game phase
  turnCount: number;
  finalTurnTriggered: boolean;

  // Coalition context (for final round)
  vintoCallerId: string | null; // Player who called Vinto
  coalitionLeaderId: string | null; // Coalition leader (if selected)

  // For evaluation
  isTerminal: boolean;
  winner: string | null;
}

/**
 * MCTS Tree Node
 */
export class MCTSNode {
  state: MCTSGameState;
  move: MCTSMove | null;
  parent: MCTSNode | null;
  children: MCTSNode[];

  visits: number;
  totalReward: number;

  untriedMoves: MCTSMove[];
  isTerminal: boolean;
  isFullyExpanded: boolean;

  constructor(
    state: MCTSGameState,
    move: MCTSMove | null,
    parent: MCTSNode | null
  ) {
    this.state = state;
    this.move = move;
    this.parent = parent;
    this.children = [];

    this.visits = 0;
    this.totalReward = 0;

    this.untriedMoves = [];
    this.isTerminal = state.isTerminal;
    this.isFullyExpanded = false;
  }

  /**
   * Get average reward (exploitation term)
   */
  getAverageReward(): number {
    return this.visits > 0 ? this.totalReward / this.visits : 0;
  }

  /**
   * Check if this node has untried moves
   */
  hasUntriedMoves(): boolean {
    return this.untriedMoves.length > 0;
  }

  /**
   * Get a random untried move
   */
  getRandomUntriedMove(): MCTSMove | null {
    if (this.untriedMoves.length === 0) return null;

    const index = Math.floor(Math.random() * this.untriedMoves.length);
    return this.untriedMoves.splice(index, 1)[0];
  }

  /**
   * Add a child node
   */
  addChild(child: MCTSNode): void {
    this.children.push(child);

    if (this.untriedMoves.length === 0) {
      this.isFullyExpanded = true;
    }
  }

  /**
   * Select best child using UCB1 formula
   */
  selectBestChildUCB1(explorationConstant: number): MCTSNode | null {
    if (this.children.length === 0) return null;

    let bestScore = -Infinity;
    let bestChild: MCTSNode | null = null;

    for (const child of this.children) {
      if (child.visits === 0) {
        // Unvisited child has infinite value
        return child;
      }

      const exploitation = child.totalReward / child.visits;
      const exploration =
        explorationConstant * Math.sqrt(Math.log(this.visits) / child.visits);

      const ucb1Score = exploitation + exploration;

      if (ucb1Score > bestScore) {
        bestScore = ucb1Score;
        bestChild = child;
      }
    }

    return bestChild;
  }

  /**
   * Select child with highest visit count (most robust choice)
   */
  selectMostVisitedChild(): MCTSNode | null {
    if (this.children.length === 0) return null;

    let maxVisits = -1;
    let bestChild: MCTSNode | null = null;

    for (const child of this.children) {
      if (child.visits > maxVisits) {
        maxVisits = child.visits;
        bestChild = child;
      }
    }

    return bestChild;
  }

  /**
   * Backpropagate reward up the tree
   */
  backpropagate(reward: number): void {
    this.visits++;
    this.totalReward += reward;

    if (this.parent) {
      this.parent.backpropagate(reward);
    }
  }

  /**
   * Get depth of this node in the tree
   */
  getDepth(): number {
    let depth = 0;
    let current: MCTSNode | null = this.parent;

    while (current !== null) {
      depth++;
      current = current.parent;
    }

    return depth;
  }

  /**
   * Debug string representation
   */
  toString(): string {
    const moveStr = this.move
      ? `${this.move.type} by ${this.move.playerId}`
      : 'ROOT';
    return `Node[${moveStr}] visits=${
      this.visits
    } reward=${this.totalReward.toFixed(2)} children=${this.children.length}`;
  }
}

/**
 * MCTS Configuration
 */
export interface MCTSConfig {
  iterations: number;
  explorationConstant: number;
  rolloutDepth: number;
  timeLimit: number; // milliseconds
}

/**
 * Default MCTS configurations by difficulty
 */
export const MCTS_DIFFICULTY_CONFIGS: Record<
  'easy' | 'moderate' | 'hard',
  MCTSConfig
> = {
  easy: {
    iterations: 500,
    explorationConstant: Math.sqrt(2),
    rolloutDepth: 5,
    timeLimit: 500,
  },
  moderate: {
    iterations: 1_500,
    explorationConstant: Math.sqrt(2),
    rolloutDepth: 10,
    timeLimit: 1_000,
  },
  hard: {
    iterations: 5_000,
    explorationConstant: Math.sqrt(2),
    rolloutDepth: 15,
    timeLimit: 1_500,
  },
};

File: kmp/shared/bot/src/commonMain/kotlin/game/vinto/bot/MctsTypes.kt
package game.vinto.bot

import game.vinto.shapes.Card
import game.vinto.shapes.Difficulty
import game.vinto.shapes.Pile
import game.vinto.shapes.Rank
import kotlin.math.ln
import kotlin.math.sqrt
import kotlin.random.Random

/**
 * The search's own model of the game, ported from `packages/bot/src/lib/mcts-types.ts`.
 *
 * This is deliberately *not* `GameState`. The bot searches over what it believes, which
 * means opponents are card counts plus memories rather than hands, and the cards it cannot
 * see are filled in by sampling (see determinization). Searching the real state would be
 * both cheating and far more expensive.
 */

enum class MctsMoveType { DRAW, TAKE_DISCARD, USE_ACTION, SWAP, DISCARD, TOSS_IN, CALL_VINTO, PASS }

data class MctsActionTarget(val playerId: String, val position: Int)

data class MctsMove(
    val type: MctsMoveType,
    val playerId: String,
    val actionCard: Card? = null,
    val targets: List<MctsActionTarget> = emptyList(),
    val swapPosition: Int? = null,
    val declaredRank: Rank? = null,
    val tossInPositions: List<Int> = emptyList(),
    val shouldSwap: Boolean? = null,
)

/** An opponent as the bot sees them: a hand size, some memories, and an estimate. */
data class MctsPlayerState(
    val id: String,
    val cardCount: Int,
    val knownCards: Map<Int, CardMemory> = emptyMap(),
    val score: Double = 0.0,
)

data class MctsGameState(
    val players: List<MctsPlayerState>,
    val currentPlayerIndex: Int,
    val botPlayerId: String,
    val discardPileTop: Card? = null,
    val discardPile: Pile = Pile(),
    val deckSize: Int = 0,
    /**
     * How many cards the model thinks are on the discard pile — the real pile at the root,
     * bookkept by [StateTransition] after that. It exists so a rollout can model the
     * reshuffle: the real game folds the pile back into the deck rather than ending, and a
     * search that could not see past an empty deck treated the endgame as a wall.
     */
    val discardCount: Int = 0,
    val botMemory: BotMemory,
    /** Cards filled in by determinization, keyed `playerId-position`. */
    val hiddenCards: Map<String, Card> = emptyMap(),
    val pendingCard: Card? = null,
    val isTossInPhase: Boolean = false,
    val tossInRanks: List<Rank> = emptyList(),
    val turnCount: Int = 0,
    val finalTurnTriggered: Boolean = false,
    val vintoCallerId: String? = null,
    val coalitionLeaderId: String? = null,
    val opponentModeler: OpponentModeler? = null,
    val isTerminal: Boolean = false,
    val winner: String? = null,
) {
    fun hiddenCardKey(playerId: String, position: Int) = "$playerId-$position"
}

/**
 * A node in the search tree.
 *
 * Mutable by nature — MCTS grows and rewrites this tree thousands of times per decision, and
 * rebuilding it immutably would dominate the cost of the search itself.
 */
class MctsNode(
    val state: MctsGameState,
    val move: MctsMove?,
    val parent: MctsNode?,
) {
    val children = mutableListOf<MctsNode>()

    var visits: Int = 0
        private set
    var totalReward: Double = 0.0
        private set

    var untriedMoves: MutableList<MctsMove> = mutableListOf()
    val isTerminal: Boolean = state.isTerminal
    var isFullyExpanded: Boolean = false
        private set

    fun averageReward(): Double = if (visits > 0) totalReward / visits else 0.0

    fun hasUntriedMoves(): Boolean = untriedMoves.isNotEmpty()

    /** Takes an untried move at random. [random] is injected so a search is reproducible. */
    fun takeRandomUntriedMove(random: Random): MctsMove? {
        if (untriedMoves.isEmpty()) return null
        return untriedMoves.removeAt(random.nextInt(untriedMoves.size))
    }

    fun addChild(child: MctsNode) {
        children += child
        if (untriedMoves.isEmpty()) isFullyExpanded = true
    }

    /**
     * UCB1: exploit what looks good, explore what is untested.
     *
     * An unvisited child is returned immediately — its value is unbounded, and trying
     * everything once before comparing is what stops the search committing to the first
     * move that happened to look reasonable.
     */
    fun selectBestChildUcb1(explorationConstant: Double): MctsNode? {
        if (children.isEmpty()) return null

        var best: MctsNode? = null
        var bestScore = Double.NEGATIVE_INFINITY

        for (child in children) {
            if (child.visits == 0) return child

            val exploitation = child.totalReward / child.visits
            val exploration = explorationConstant * sqrt(ln(visits.toDouble()) / child.visits)
            val score = exploitation + exploration

            if (score > bestScore) {
                bestScore = score
                best = child
            }
        }
        return best
    }

    /**
     * The move to actually play: the most *visited* child, not the best-scoring one.
     *
     * A high average from few visits is usually luck; visit count is what the search
     * committed its time to, which makes it the more robust choice.
     */
    fun selectMostVisitedChild(): MctsNode? = children.maxByOrNull { it.visits }

    fun backpropagate(reward: Double) {
        var node: MctsNode? = this
        while (node != null) {
            node.visits++
            node.totalReward += reward
            node = node.parent
        }
    }

    fun depth(): Int {
        var depth = 0
        var current = parent
        while (current != null) {
            depth++
            current = current.parent
        }
        return depth
    }
}

/**
 * How hard the bot thinks.
 *
 * `timeLimitMillis` is a safety valve for a live game, and is deliberately **null by
 * default**: a search that stops on the clock produces different moves on different
 * machines, which would make a recorded game unreplayable. Deterministic runs — tests,
 * replays, the parity corpus — use the iteration budget alone (design D4).
 */
data class MctsConfig(
    val iterations: Int,
    val explorationConstant: Double,
    val rolloutDepth: Int,
    val timeLimitMillis: Long? = null,
)

val MCTS_DIFFICULTY_CONFIGS: Map<Difficulty, MctsConfig> = mapOf(
    Difficulty.EASY to MctsConfig(iterations = 500, explorationConstant = 1.8, rolloutDepth = 15),
    Difficulty.MODERATE to MctsConfig(iterations = 2_000, explorationConstant = 1.6, rolloutDepth = 20),
    Difficulty.HARD to MctsConfig(iterations = 5_000, explorationConstant = 1.4, rolloutDepth = 30),
)
